pkg/client/kubernetes/cache/single_object.go

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener/pkg/utils"
+)
+
+var _ cache.Cache = &singleObject{}
+
+type singleObject struct {
+	parentCtx  context.Context
+	log        logr.Logger
+	restConfig *rest.Config
+	newCache   cache.NewCacheFunc
+	opts       func() cache.Options
+
+	started   bool
+	startWait chan struct{} // startWait is a channel that is closed after the cache has been started
+
+	lock  sync.RWMutex
+	store map[client.ObjectKey]*objectCache
+
+	clock                     clock.Clock
+	garbageCollectionInterval time.Duration
+	maxIdleTime               time.Duration
+}
+
+type objectCache struct {
+	cache          cache.Cache
+	cancel         context.CancelFunc
+	lastAccessTime atomic.Value
+}
+
+// NewSingleObject creates a new instance of the singleObject cache.Cache implementation.
+// This cache maintains a separate cache per `client.ObjectKey` and invalidates them when not accessed for the
+// given `maxIdleTime`. A new cache for a particular object is added or re-added as soon as the caches `Get()` function
+// is invoked. Please note that object types are not differentiated by this cache (only object keys), i.e. it must not
+// be used with mixed GVKs.
```

Comment:
Wouldn't it make sense to overcome the limitation regarding mixed GVKs by creating a "nested" `singleObject.store` like `map[schema.GroupVersionKind]{current store definition}`?
Changing that does not look very complex.

Reply:
Yes, this would be possible, but I'm not sure whether we should go ahead and change this now after the implementation is done and tested. I don't think it's a big limitation.

Introducing such bug where the same cache is used for multiple GVK is also quite unlikely, I believe: You would need to create a single object cache and then reference the it [here](https://github.com/gardener/gardener/pull/7632/files#diff-eb33ba906d976101dfe4f69a634d451fd48618d3bfdf57fe7401aca9e03b9bd8R311-R320) for multiple resources (despite the doc string explicitly stating that this is not supported) instead of creating a dedicated cache like it's done now.

I would propose to go ahead as is, get some experiences, and come back if we see the necessity.

Reply:
I also think that the benefit of this is minimal because a user of the `SingleObject` cache will probably always only use it for a dedicated `GVK`s, so separation is required at a different layer, .e.g. [aggregated cache](https://github.com/gardener/gardener/blob/b2d32fe231cb47926480eb30a1d8f2ca6b9be4dc/cmd/gardenlet/app/app.go#L320).